plasmapy/tests/helpers/exceptions.py

```diff
@@ -0,0 +1,105 @@
+"""Exceptions that describe different types of test failures."""
+
+__all__ = [
+    "ExceptionMismatchFail",
+    "InvalidTestError",
+    "MissingExceptionFail",
+    "MissingWarningFail",
+    "TestFailure",
+    "TypeMismatchFail",
+    "UnexpectedExceptionFail",
+    "UnexpectedResultFail",
+    "UnexpectedWarningFail",
+    "WarningMismatchFail",
+]
+
+import pytest
+
+
+class TestFailure(pytest.fail.Exception):
```

Comment:
This is the first time I've looked into `pytest`'s file structure and man is it weird!

Using `pytest.fail.Exception` is very round about way of getting the `Exception`.  It's not a literal exception definition, but a class attribute of `_WithException` which is used in decorating `pytest.fail()`.  Instead, you should use `_pytest.outcomes.Failed`, which equates to the parent class of `pytest.fail.Excpetion`.

Reply:
You'll enjoy this one...How about `TestFailed` instead of `TestFailure` to better mirror `_pytest.outcome.Failed`?  That is all I will say and I'll let you choose whichever you prefer. 😆

Reply:
I had thought about that, but was hesitant to because it didn't seem like part of the public API.  I'm also not sure if it adds much to inherit from the pytest exception.  In any case, I'm okay with any of these possibilities.

And yeah...the inside of pytest is pretty different from what I'm used to.

Reply:
> You'll enjoy this one...How about `TestFailed` instead of `TestFailure` to better mirror `_pytest.outcome.Failed`? That is all I will say and I'll let you choose whichever you prefer. laughing

This I'll agree with you on!

Reply:
> Instead, you should use `_pytest.outcomes.Failed`, which equates to the parent class of `pytest.fail.Excpetion`.

Hm...it appears that `pytest._pytest` cannot be imported.  I think that means we'll need to go with `pytest.fail.Exception` unless I'm missing something.

```Python
In [8]: import pytest._pytest
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-8-148f78f1fe80> in <module>
----> 1 import pytest._pytest

ModuleNotFoundError: No module named 'pytest._pytest'
```

Reply:
> Shall I go ahead with inheriting from _pytest.outcomes.Failed, given the uncertainty about whether or not it's part of the public API?

I'd say yes.  I think that approach would be more stable than the other route.  My only concern is that by sub-classing `_pytest.outcomes.Failed` our custom failures might start interacting with the internals of `pytest`.  If we find that this happens in an undesirable way, then we make our own `Failed` class that inherits from `BaseException` or `Exception`.

Reply:
You should do...

```python
from _pytest.outcomes import Failed

class TestFailed(Failed):
   ...
```

`_pytest` is installed with `pytest` because `pytest` basically calls on `_pytest`.